swag/bank.py

```diff
@@ -301,6 +308,226 @@ def set_timezone(self, user, timezone):
+            cagnotte
+            for cagnotte in self.swagdb.get_cagnotte_infos()
+            if cagnotte.cagnotte_activation == True
+        ]
+
+    def get_cagnotte_history(self, cagnotte_idx):
+        return [
+            transaction
+            for transaction in self.swagdb.blockchain
+            if concerns_cagnotte(cagnotte_idx, transaction)
+        ]
+
+    def payer_a_cagnotte(
+        self, donator_account_discord_id: int, cagnotte_idx: int, montant
+    ):
+        cagnotte = self.get_active_cagnotte(cagnotte_idx)
+        donator_account = self.swagdb.get_account(donator_account_discord_id)
+        # On regarde le type de la cagnotte pour pouvoir correctement choisir les fonctions qui devront être utiliser
+        if cagnotte.get_info().cagnotte_currency == "$wag":
+
+            # Check if the value of $wag is correct regardless its propriety
+            if not isinstance(montant, int) or montant < 0:
+                raise InvalidSwagValue
+
+            # Check if the donator have enough $wag:
+            if (
+                self.get_account_info(donator_account_discord_id).swag_balance - montant
+                < 0
+            ):
+                raise NotEnoughSwagInBalance(donator_account_discord_id)
+
+            # Making the donation
+            donator_account.swag_balance -= montant
+            cagnotte.cagnotte_balance += montant
+
+        elif cagnotte.get_info().cagnotte_currency == "$tyle":
+
+            # Check if the value of $tyle is correct regardless its propriety
+            if not isinstance(montant, (int, float, Decimal)) or montant < 0:
+                raise InvalidStyleValue
+
+            # Check if the donator have enough $tyle:
+            if (
+                self.get_account_info(donator_account_discord_id).style_balance
+                - montant
+                < 0
+            ):
+                raise NotEnoughStyleInBalance
+
+            donator_account.style_balance -= montant
+            cagnotte.cagnotte_balance += montant
+
+        # Write donation in the blockchain
+        self.swagdb.blockchain.append(
+            (
+                utcnow().datetime,
+                TransactionType.DONATION,
+                (
+                    donator_account.get_info().id,
+                    cagnotte.get_info().cagnotte_id,
+                    montant,
+                    cagnotte.get_info().cagnotte_currency,
+                ),
+            )
+        )
+
+        if donator_account_discord_id not in cagnotte.get_info().cagnotte_participant:
+            cagnotte.cagnotte_participant.append(
+                donator_account_discord_id
+            )  # TODO voir si ça marche
+
+        self.transactional_save()
+
+    def donner_depuis_cagnotte(
+        self,
+        cagnotte_idx: int,
+        receiver_account_discord_id: int,
+        montant,
+        emiter_account_discord_id: int,
+    ):
+        cagnotte = self.get_active_cagnotte(cagnotte_idx)
```

Comment:
Because `get_active_cagnotte` only return Cagnotte if it's active. If not, the function return an exception, that's why I prefered to precise in the name **active**.

On the other hand, if we name it just `get_cagnotte` I'm afraid we could possibly misunderstood(stand?) its behavior.

Reply:
Yeah, but since deactivated Cagnottes are unreachable, does the term "active Cagnotte" bear meaning? Do we need to get deactivated Cagnotte for some purpose? I feel like the very fact that this function throw a "Cagnotte inactive" exception is hint that it's not the case **but** therefore we can ask ourselves how the blockchain can access deactivated Cagnottes infos

Reply:
Indeed, in the futur maybe we could imagine ressurect old cagnotte or simply want to display a desactivated €agnotte in history